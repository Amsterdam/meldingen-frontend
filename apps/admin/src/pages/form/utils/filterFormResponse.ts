import type { FormInput } from '@meldingen/api-client'
import {
  FormCheckboxComponentInputSchema,
  FormComponentInputValidateSchema,
  FormInputSchema,
  FormPanelComponentInputSchema,
  FormRadioComponentInputSchema,
  FormSelectComponentInputSchema,
  FormTextAreaComponentInputSchema,
  FormTextFieldComponentInputSchema,
} from '@meldingen/api-client'
import filter from 'uber-json-schema-filter'

const filterBySchemaPerType = (obj: any) => {
  switch (obj.type) {
    case 'panel':
      return filter(FormPanelComponentInputSchema, obj)
    case 'radio':
      return filter(FormRadioComponentInputSchema, obj)
    case 'select':
      return filter(FormSelectComponentInputSchema, obj)
    case 'selectboxes':
      return filter(FormCheckboxComponentInputSchema, obj)
    case 'textarea':
      // Add autoExpand to object if it doesn't exist, the builder doesn't do that by default
      if (!Object.hasOwn(obj, 'autoExpand')) {
        return filter(FormTextAreaComponentInputSchema, { ...obj, autoExpand: false })
      }

      return filter(FormTextAreaComponentInputSchema, obj)
    case 'textfield':
      return filter(FormTextFieldComponentInputSchema, obj)
    default:
      throw Error(`Type ${obj.type} is unknown, please add it to filterFormResponse.`)
  }
}

const getFilteredValidateObject = (validateObj: any) => {
  const validate = filter(FormComponentInputValidateSchema, validateObj)

  // Explicitly remove the 'json' key if its value is an empty string, the API doesn't accept that
  if (validate?.json === '') {
    delete validate.json
  }

  return validate
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const filterFormResponse = (obj: any): FormInput => {
  // This function is used to filter an additional questions form, which is always
  // a wizard with panels containing questions. Therefore the form has a fixed depth of two levels.
  const firstLevelComponents = obj.components.map((firstLevelComponent: any) => {
    const secondLevelComponents = firstLevelComponent.components.map((secondLevelComponent: any) => ({
      ...filterBySchemaPerType(secondLevelComponent),
      validate: getFilteredValidateObject(secondLevelComponent.validate),
    }))

    const filteredObject = {
      ...filterBySchemaPerType(firstLevelComponent),
      components: secondLevelComponents,
      validate: getFilteredValidateObject(firstLevelComponent.validate),
    }

    // Explicitly remove the 'validate' key for panel components, the API doesn't accept that
    if (filteredObject.type === 'panel') {
      delete filteredObject.validate
    }

    return filteredObject
  })

  return {
    ...filter(FormInputSchema, obj),
    components: firstLevelComponents,
  }
}
