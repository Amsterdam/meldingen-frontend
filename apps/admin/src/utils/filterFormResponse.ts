const allowLists: Record<string, string[]> = {
  panel: ['label', 'key', 'type', 'input', 'components'],
  textarea: ['label', 'description', 'key', 'type', 'input', 'autoExpand', 'showCharCount'],
}

// TODO: second 'any' should be a type supplied by the BE?
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const filterFormResponse = (obj: any): any => {
  // If the input is a JS primitive, return the unmodified input
  if (obj !== Object(obj)) {
    return obj
  }

  // If the input is an array, run this function recursively on all items of the array
  if (Array.isArray(obj)) {
    return obj.map((item) => filterFormResponse(item))
  }

  return (
    Object.keys(obj)
      // If input has a key called 'type', and the value of 'type' is present in the keys of the allowLists object,
      // filter keys based on that specific allow list. Otherwise, all keys are passed.
      .filter((key) => {
        if (Object.keys(allowLists).includes(obj?.type)) {
          return allowLists[obj.type].includes(key)
        }

        return true
      })
      // Reduce the input by recursively calling the function on each value and assigning it to the key
      .reduce((acc, x) => Object.assign(acc, { [x]: filterFormResponse(obj[x]) }), {})
  )
}
