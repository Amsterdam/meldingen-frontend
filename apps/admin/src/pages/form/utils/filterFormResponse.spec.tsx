import { filterFormResponse } from './filterFormResponse'
import {
  rootObjMock,
  panelsMockBefore,
  panelsMockAfter,
  radiosMockBefore,
  radiosMockAfter,
  selectsMockBefore,
  selectsMockAfter,
  textAreasMockBefore,
  textAreasMockAfter,
  textFieldMockBefore,
  textFieldMockAfter,
} from './filterFormResponseMocks'

describe('filterFormResponse', () => {
  it('should not filter the root object', () => {
    expect(filterFormResponse(rootObjMock)).toEqual(rootObjMock)
  })

  it('should filter the panel components', () => {
    const before = { ...rootObjMock, components: panelsMockBefore }
    const after = { ...rootObjMock, components: panelsMockAfter }

    expect(filterFormResponse(before)).toEqual(after)
  })

  it('should filter the radio components', () => {
    const before = { ...rootObjMock, components: { ...panelsMockBefore, components: radiosMockBefore } }
    const after = { ...rootObjMock, components: { ...panelsMockAfter, components: radiosMockAfter } }

    expect(filterFormResponse(before)).toEqual(after)
  })

  it('should filter the select components', () => {
    const before = { ...rootObjMock, components: { ...panelsMockBefore, components: selectsMockBefore } }
    const after = { ...rootObjMock, components: { ...panelsMockAfter, components: selectsMockAfter } }

    expect(filterFormResponse(before)).toEqual(after)
  })

  it('should filter the text area components', () => {
    const before = { ...rootObjMock, components: { ...panelsMockBefore, components: textAreasMockBefore } }
    const after = { ...rootObjMock, components: { ...panelsMockAfter, components: textAreasMockAfter } }

    expect(filterFormResponse(before)).toEqual(after)
  })

  it('should filter the text field components', () => {
    const before = { ...rootObjMock, components: { ...panelsMockBefore, components: textFieldMockBefore } }
    const after = { ...rootObjMock, components: { ...panelsMockAfter, components: textFieldMockAfter } }

    expect(filterFormResponse(before)).toEqual(after)
  })
})
