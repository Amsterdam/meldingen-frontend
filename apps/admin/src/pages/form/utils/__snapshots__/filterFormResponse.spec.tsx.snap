// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filterFormResponse renders the correct output 1`] = `
{
  "classification": null,
  "components": [
    {
      "components": [
        {
          "input": true,
          "key": "textField",
          "label": "Text Field",
          "type": "textfield",
          "validate": undefined,
        },
        {
          "autoExpand": false,
          "input": true,
          "key": "textArea",
          "label": "Text Area",
          "type": "textarea",
          "validate": undefined,
        },
        {
          "input": true,
          "key": "selectBoxes",
          "label": "Select Boxes",
          "type": "selectboxes",
          "validate": undefined,
          "values": [
            {
              "label": "one",
              "shortcut": "",
              "value": "one",
            },
            {
              "label": "two",
              "shortcut": "",
              "value": "two",
            },
          ],
        },
        {
          "data": {
            "values": [
              {
                "label": "one",
                "value": "one",
              },
              {
                "label": "two",
                "value": "two",
              },
            ],
          },
          "input": true,
          "key": "select",
          "label": "Select",
          "type": "select",
          "validate": undefined,
          "widget": "html5",
        },
        {
          "input": true,
          "key": "radio",
          "label": "Radio",
          "type": "radio",
          "validate": undefined,
          "values": [
            {
              "label": "one",
              "shortcut": "",
              "value": "one",
            },
            {
              "label": "two",
              "shortcut": "",
              "value": "two",
            },
          ],
        },
      ],
      "input": false,
      "key": "page1",
      "label": "Page 1",
      "type": "panel",
    },
  ],
  "display": "wizard",
  "title": "Test title",
}
`;
